### PR TITLE
Interpolant production and iterativeness for lookahead

### PR DESCRIPTION
**Pull request by BritikovKI**
Another algorithm for SMT solving.
Comparing to the lookahead it allows to have a flexibility, as it doesn't run on all of the literals in the instance.
This algoritm is significantly more performant then lookahead and allows a middle ground.
Depending on the configuration it can actually be executed as just CoreSMT solver or Lookahead.

experiments_vsids_vs_picky

experiments_lookahead_vs_picky

### CODE REVIEW

File: src/smtsolvers/LookaheadSMTSolver.cc [hunk 1]
```diff
@@ -76,18 +78,14 @@ lbool LookaheadSMTSolver::laPropagateWrapper() {
             if (decisionLevel() == 0)
                 return l_False; // Unsat
             -- confl_quota;
-#ifdef LADEBUG
-            cerr << "; Got a conflict, quota now " << confl_quota.getQuota() << "\n";
-#endif
+            // Received a conflict ad decisionLevel > 0
```
Comment (blishko) [inline src/smtsolvers/LookaheadSMTSolver.cc:81]:
```suggestion
            // Received a conflict and decisionLevel > 0
```

File: src/smtsolvers/LookaheadSMTSolver.cc [hunk 2]
```diff
@@ -147,43 +134,46 @@ LookaheadSMTSolver::PathBuildResult LookaheadSMTSolver::setSolverToNode(LANode c
...
-            } else if (res == l_Undef) {
-                cancelUntil(0);
-                return PathBuildResult::pathbuild_restart; // Do a restart
-            }
-            if (curr_dl != decisionLevel()) {
-                return PathBuildResult::pathbuild_unsat;
-            }
+    // setting solver to the correct dl
+    int i = 0;
+    if(path.size() <= decisionLevel()) {
```
Comment (blishko) [inline src/smtsolvers/LookaheadSMTSolver.cc:139]: Spacing

```suggestion
    if (path.size() <= decisionLevel()) {
```

File: src/smtsolvers/LookaheadSMTSolver.cc [hunk 3]
```diff
@@ -147,43 +134,46 @@ LookaheadSMTSolver::PathBuildResult LookaheadSMTSolver::setSolverToNode(LANode c
...
+
+        if (path.size() > 0) { cancelUntil(path.size() - 1);}
+        if (path.size() == 0) { cancelUntil(0);}
```
Comment (blishko) [inline src/smtsolvers/LookaheadSMTSolver.cc:142]:
```suggestion
        if (path.size() > 0) { cancelUntil(path.size() - 1); }
        else { cancelUntil(0); }
```

File: src/smtsolvers/LookaheadSMTSolver.cc [hunk 4]
```diff
@@ -147,43 +134,46 @@ LookaheadSMTSolver::PathBuildResult LookaheadSMTSolver::setSolverToNode(LANode c
...
-            printf("Would propagate %s%d but the literal is already assigned\n", sign(path[i]) ? "-" : "", var(path[i]));
-#endif
-            if (value(path[i]) == l_False) {
-#ifdef LADEBUG
-                printf("Unsatisfiable branch since I'd like to propagate %s%d but %s%d is assigned already\n", sign(path[i]) ? "-" : "", var(path[i]), sign(~path[i]) ? "-" : "", var(path[i]));
-                printf("Marking the subtree false:\n");
-                n->print();
-#endif
-                return PathBuildResult::pathbuild_unsat;
+            crossed_assumptions = path.size();
```
Comment (blishko) [inline src/smtsolvers/LookaheadSMTSolver.cc:146]: crossed_assumptions = std::min(path.size(), assumptions.size());

File: src/smtsolvers/LookaheadSMTSolver.cc [hunk 5]
```diff
@@ -147,43 +134,46 @@ LookaheadSMTSolver::PathBuildResult LookaheadSMTSolver::setSolverToNode(LANode c
...
+        }
+    } else {
+        i = path.size() - decisionLevel() - 1;
```
Comment (blishko) [inline src/smtsolvers/LookaheadSMTSolver.cc:149]: Please add parentheses for clarification.

File: src/smtsolvers/LookaheadSMTSolver.cc [hunk 6]
```diff
@@ -147,43 +134,46 @@ LookaheadSMTSolver::PathBuildResult LookaheadSMTSolver::setSolverToNode(LANode c
...
-                assert(value(path[i]) == l_True);
+                // literal to propagate was already assigned
+                if (value(path[i]) == l_False) {
+                    return PathBuildResult::pathbuild_unsat;
+                } else {
+                    assert(value(path[i]) == l_True);
+                }
             }
         }
+        rebuildOrderHeap();
```
Comment (blishko) [inline src/smtsolvers/LookaheadSMTSolver.cc:176]: Why is this necessary now?

File: src/smtsolvers/LookaheadSMTSolver.cc [hunk 7]
```diff
@@ -239,16 +227,12 @@ std::pair<LookaheadSMTSolver::laresult,Lit> LookaheadSMTSolver::lookaheadLoop()
     bool respect_logic_partitioning_hints = config.respect_logic_partitioning_hints();
     int skipped_vars_due_to_logic = 0;
 
-#ifdef LADEBUG
-    printf("Starting lookahead loop with %d vars\n", nVars());
-#endif
+    Lit best;
```
Comment (blishko) [inline src/smtsolvers/LookaheadSMTSolver.cc:230]: Now there are two `best` declaration in this function. I think the previous version was fine.

File: src/smtsolvers/LookaheadSMTSolver.cc [hunk 8]
```diff
@@ -342,28 +309,47 @@ std::pair<LookaheadSMTSolver::laresult,Lit> LookaheadSMTSolver::lookaheadLoop()
...
-#endif
+
+    // lookahead phase is over
     idx = (idx + i) % nVars();
     if (!okToPartition(var(best))) { unadvised_splits++; }
     return {laresult::la_ok, best};
 }
+
+
+void LookaheadSMTSolver::cancelUntil(int level)
```
Comment (blishko) [inline src/smtsolvers/LookaheadSMTSolver.cc:330]: This seems mostly like a duplicate of ``CoreSMTSolver::cancelUntil, with only one line added.

I think you could do the following, no?
```
void LookaheadSMTSolver::cancelUntil(int level) {
CoreSMTSolver::cancelUntil(level);
crossed_assumptions = min(crossed_assumptions, level);
}
```

File: src/smtsolvers/LookaheadSMTSolver.h [hunk 9]
```diff
@@ -151,27 +153,57 @@ LookaheadSMTSolver::buildAndTraverse(BuildConfig && buildConfig) {
...
+                    crossed_assumptions++;
+                    checked = true;
+                    queue.push(c1_raw);
+                    break;
+                }
+            }
+        }
 
-        switch (expandTree(*n, std::move(c1), std::move(c2))) {
+        if(!checked) {
```
Comment (blishko) [inline src/smtsolvers/LookaheadSMTSolver.h:189]: Spacing.

File: src/smtsolvers/LookaheadSMTSolver.cc [hunk 10]
```diff
@@ -42,17 +44,17 @@ lbool LookaheadSMTSolver::solve_() {
...
-        case LALoopRes::sat:
-            return l_True;
-        case LALoopRes::unsat: {
-            ok = false;
-            return l_False;
-        }
-        default:
-            assert(false);
-            return l_Undef;
+    case LALoopRes::unknown_final:
```
Comment (blishko) [inline src/smtsolvers/LookaheadSMTSolver.cc:47]: The new indentation seems wrong. I think the previous one is correct.

**Review by blishko — COMMENTED** [review]
I left some questions and suggestions.

File: src/smtsolvers/LookaheadSMTSolver.cc (hunk 2 above)
Comment (BritikovKI) [reply src/smtsolvers/LookaheadSMTSolver.cc]: Fixed

File: src/smtsolvers/LookaheadSMTSolver.cc (hunk 3 above)
Comment (BritikovKI) [reply src/smtsolvers/LookaheadSMTSolver.cc]: Fixed

File: src/smtsolvers/LookaheadSMTSolver.cc (hunk 4 above)
Comment (BritikovKI) [reply src/smtsolvers/LookaheadSMTSolver.cc]: Fixed

File: src/smtsolvers/LookaheadSMTSolver.cc (hunk 5 above)
Comment (BritikovKI) [reply src/smtsolvers/LookaheadSMTSolver.cc:157]: Fixed

File: src/smtsolvers/LookaheadSMTSolver.cc (hunk 6 above)
Comment (BritikovKI) [reply src/smtsolvers/LookaheadSMTSolver.cc:184]: It is necessary because we could've added new clauses, so the score of the literals may've changed

File: src/smtsolvers/LookaheadSMTSolver.cc (hunk 7 above)
Comment (BritikovKI) [reply src/smtsolvers/LookaheadSMTSolver.cc:238]: Fixed

File: src/smtsolvers/LookaheadSMTSolver.cc (hunk 8 above)
Comment (BritikovKI) [reply src/smtsolvers/LookaheadSMTSolver.cc:338]: Fixed

File: src/smtsolvers/LookaheadSMTSolver.h (hunk 9 above)
Comment (BritikovKI) [reply src/smtsolvers/LookaheadSMTSolver.h]: Fixed

File: src/smtsolvers/LookaheadSMTSolver.cc (hunk 10 above)
Comment (BritikovKI) [reply src/smtsolvers/LookaheadSMTSolver.cc]: Changed

File: src/smtsolvers/LookaheadSMTSolver.cc (hunk 5 above)
Comment (blishko) [reply src/smtsolvers/LookaheadSMTSolver.cc:157]: I meant add parentheses to make it clear if this is `(path.size() - decisionLevel()) - 1` or `path.size() - (decisionLevel() - 1)`

File: src/smtsolvers/LookaheadSMTSolver.cc [hunk 11]
```diff
@@ -147,43 +143,45 @@ LookaheadSMTSolver::PathBuildResult LookaheadSMTSolver::setSolverToNode(LANode c
...
+        // New path is different from the old one only in the last element
+        if (path.size() > 0) { cancelUntil(path.size() - 1); }
+        else { cancelUntil(0); }
+        crossed_assumptions = std::min(path.size(), assumptions.size());
+    } else {
+        // Means no conflict was encountered and basic path hasn't changed
+        // we need only propagate i new literals
+        i = path.size() - decisionLevel() - 1;
+    }
+    if(path.size() > 0){
```
Comment (blishko) [inline src/smtsolvers/LookaheadSMTSolver.cc:159]: Spacing. Please run formatter on the code.

Comment (BritikovKI) [reply src/smtsolvers/LookaheadSMTSolver.cc]: Sorry for that, I ran the formatter, I think I should maybe do some script to make it a pre-commit hook, so I don't forget it

File: src/smtsolvers/LookaheadSMTSolver.h [hunk 12]
```diff
@@ -111,69 +92,99 @@ LookaheadSMTSolver::buildAndTraverse(BuildConfig && buildConfig) {
     auto root = std::unique_ptr<Node>(root_raw);
     root->p = root_raw;
     queue.push(root_raw);
+    TPropRes res = checkTheory(true);
+    if (res == TPropRes::Unsat) {
+        return {LALoopRes::unsat, nullptr};
+        ; // Unsat
```
Comment (blishko) [inline src/smtsolvers/LookaheadSMTSolver.h:98]: This line should probably be removed.

**Review by blishko — APPROVED** [review]
OK! Next time, we will work on the commit messages :)
